nixos/doc/manual/development/writing-nixos-tests.section.md

```diff
@@ -415,3 +475,120 @@ import ./make-test-python.nix
 ```
 
 In that case, `numpy` is chosen from the generic `python3Packages`.
+
+## Constructing a test matrix {#sec-nixos-test-matrix}
+
+You may want to run variations of your test, or perhaps split long tests into
+a few shorter, focused tests.
+
+While you could roll your own solution with `mapAttrs`, this is actually
+non-trivial because of the requirements imposed by `nix-build`, hydra and
+`passthru.tests`. So instead, we recommend to use the `matrix` test options that
+take care of this for you.
+
+These options work by duplicating the entire test module configuration below
+the `matrix.<decision>.choice.<choice>.module` options.
+Laziness makes this pattern efficient.
+
+The test framework ([`runTest`](#sec-calling-nixos-tests)) takes care of
+traversing the choices.
+
+A parameterized test may look as follows:
+
+```nix
+{ lib, backend, ... }: {
+  name = "foo-${backend}";
+
+  matrix.backend.choice.smtp.module = {};
+  matrix.backend.choice.rest.module = {};
+
+  defaults.services.foo.backend = backend;
+
+  nodes = …;
+  testScript = ''
+    ${lib.optionalString (backend == "rest") ''
+      # extra setup for the rest backend
+    ''}
+    # …
+  '';
+}
+```
+
+Instead of a single derivation, the test framework now returns:
+
+```nix
+{
+  backend-rest = «derivation /nix/store/…-foo-rest.drv»;
+  backend-smtp = «derivation /nix/store/…-foo-smtp.drv»;
+
+  # make nix-build traverse this attrset
+  recurseForDerivations = true;
+  # …
+}
+```
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
+
+The test related options such as `nodes` at the root of the test configuration still exist there, but are not demanded by `runTest`, so only their copies at each of the choices are evaluated. Hence, the "root" can see the consequences of the choices.
+
+Some variations aren't simply parametric, but require specific configuration for some choices.
+These can be added in the `module` submodule:
+
+```nix
+{ lib, setup, ... }: {
+  matrix.backend.choice.smtp.module = { lib, setup, ... }: {
+    defaults.services.foo.backend = "smtp";
+    nodes.smtpserver = { … };
+    _module.args.setup = "";
+  };
+  matrix.backend.choice.rest.module = { lib, setup, ... }: {
+    _module.args.setup = ''
+      # extra setup for the rest backend
+    '';
+  };
+
+  nodes = …;
+  testScript = ''
+    ${setup}
+    # …
+  '';
+}
+```
+
+And finally, some tests truly form a matrix, as they have two (or more) parameters, such as this example, producing the 6 test derivations that form the Cartesian product of `backend` and `testsuite`.
+
+Note that the modules inside `module` submodule can access their parent's module
+arguments through the lexical scope, where the matrix choice has not been decided
+yet. Any such parameters must be ignored or repeated inside the `module` submodule
+to get the final value.
```

Comment:
I really don't understand what this means. How exactly can the `module` submodules access their parent's argument? How would one get the final value? If this is important, please expand this note into a full example.